### PR TITLE
 LINK-206, LINK-207, LINK-244, LINK-246, LINK-247 | Dropdown component which contains all action buttons for single event

### PR DESCRIPTION
**Pull request by jorilindell**
## Description
Show sub-events also in event card view
Show event hierarchy at the end of event edit page
Generic dropdown component
Replace separate action buttons with single dropdown on edit event page
Add actions dropdown for event card and event table row
Implement coping an event as template

## Closes 
[LINK-206](https://helsinkisolutionoffice.atlassian.net/browse/LINK-206)
[LINK-207](https://helsinkisolutionoffice.atlassian.net/browse/LINK-207)
[LINK-244](https://helsinkisolutionoffice.atlassian.net/browse/LINK-244)
[LINK-246](https://helsinkisolutionoffice.atlassian.net/browse/LINK-246)
[LINK-247](https://helsinkisolutionoffice.atlassian.net/browse/LINK-247)

## Screenshots
Screenshot 2021-02-08 at 15 00 01

Screenshot 2021-02-08 at 15 00 12

Screenshot 2021-02-08 at 15 00 28

Screenshot 2021-02-08 at 15 00 38

Screenshot 2021-02-08 at 15 00 59

## Testing
1. Remove comment from line 116 `(apikey: '50381be7-fef2-4783-b181-3181f6492f3f',)` on domain/app/apollo/apolloClient.ts file to use apikey for authentication 
2. Make sure that right url is set for LE by setting `REACT_APP_LINKED_EVENTS_URL=http://localhost:8000/v1` in .env.development.local
2. Setup local LE backend and set apikey for some data source
- Clone LE backend repository from https://github.com/City-of-Helsinki/linkedevents
- Add a local_settings.py file and add following lines to it to support apikey in request header
```
DEBUG = True
CORS_ALLOW_HEADERS = [
    'apikey',
    'authorization',
    'content-type',
    'dnt',
    'origin',
    'user-agent',
    'x-csrftoken',
    'x-requested-with',
]
```
- Start docker container: `docker-compose up`. Make sure ports 5432 or 8000 are not already in use
- In other terminal tab open django bash to run importers: `docker-compose exec django bash`
- At django bask run importer scripts:
```
python manage.py event_import yso --all
python manage.py add_helsinki_audience
python manage.py add_helsinki_topics
python manage.py event_import tprek --places
python manage.py import_organizations https://api.hel.fi/paatos/v1/organization/ -s helsinki:ahjo
```
- Add apikey `50381be7-fef2-4783-b181-3181f6492f3f` to one of the datasource .e.g http://localhost:8000/admin/events/datasource/helsinki/change/
Screenshot 2021-01-12 at 12 45 54

### CODE REVIEW

**Comment by Sartonon** [comment]
Screenshot 2021-02-15 at 9 06 29

Couple test failing locally for me

**Comment by jorilindell** [comment]
> Screenshot 2021-02-15 at 9 06 29
> 
> Couple test failing locally for me

Strange, I'm quite sure that those tests passed when I previous time tested them locally. But now I was also able to reproduce this. Anyway, fixed.

**Comment by Sartonon** [comment]
Screenshot 2021-02-17 at 22 57 31

Getting this error when running all tests. When I press f to run failed test, it passes

File: src/domain/events/actionsDropdown/ActionsDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    cancelEvent({});
+  };
+
+  const onDelete = async () => {
+    deleteEvent({});
+  };
+
+  const onPostpone = async () => {
+    postponeEvent({});
+  };
```
Comment (Sartonon) [inline src/domain/events/actionsDropdown/ActionsDropdown.tsx:62]: Should no parameter allowed too?

**Review by Sartonon — COMMENTED** [review]
Looking very good as always. Added couple comments

File: src/domain/events/utils.ts [hunk 2]
```diff
@@ -131,3 +140,7 @@ export const getEventsQuerySkip = (
       return !adminOrganizations.length;
   }
 };
+
+export const resetEventListPage = async () => {
+  await store.dispatch(setEventListOptions({ page: 1 }));
+};
```
[the remark this answers is not in the document]
Comment (jorilindell) [reply src/domain/events/utils.ts]: Removed extra async/await

File: src/domain/event/CreateEventPage.tsx [hunk 3]
```diff
@@ -112,6 +112,8 @@ const CreateEventPage: React.FC = () => {
 
         // Clear all events queries from apollo cache to show added events in event list
         clearEventsQueries(apolloClient);
+        // This action will change LE response so clear event list page
+        await resetEventListPage();
```
[the remark this answers is not in the document]
Comment (jorilindell) [reply src/domain/event/CreateEventPage.tsx]: Removed extra async/await

File: src/domain/event/EditEventPage.tsx [hunk 4]
```diff
@@ -67,202 +60,78 @@ interface EditEventPageProps {
...
-      // eslint-disable-next-line no-console
-      console.error(e);
-      setSaving(null);
-    }
+  const onPostpone = async () => {
+    postponeEvent({
+      onSuccess: async () => {
+        await refetch();
+        window.scrollTo(0, 0);
+      },
```
[the remark this answers is not in the document]
Comment (jorilindell) [reply src/domain/event/EditEventPage.tsx:126]: This was actually mistake in my TS type: Changed it to 
```
refetch: (
    variables?: Partial<EventQueryVariables>
  ) => Promise<ApolloQueryResult<EventQuery>>;
  ```

File: src/domain/event/hooks/useEventUpdateActions.ts [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        id: item.id,
+      }));
+
+      await updateEvents(payload);
+
+      // Call callback function if defined
+      /* istanbul ignore else */
+      if (onSuccess) {
+        await onSuccess();
+      }
```
[the remark this answers is not in the document]
Comment (jorilindell) [reply src/domain/event/hooks/useEventUpdateActions.ts]: Changed

File: src/domain/events/actionsDropdown/ActionsDropdown.tsx (hunk 1 above)
Comment (jorilindell) [reply src/domain/events/actionsDropdown/ActionsDropdown.tsx]: Yes we should. Added that possibility

File: src/domain/event/EditEventPage.tsx [hunk 6]
```diff
@@ -67,202 +60,78 @@ interface EditEventPageProps {
...
-      // Network errors will be handled on apolloClient error link. Only show error on console here.
-      // eslint-disable-next-line no-console
-      console.error(e);
-      setSaving(null);
-    }
+  const onDelete = async () => {
+    deleteEvent({
+      onSuccess: async () => {
+        // This action will change LE response so clear event list page
+        await resetEventListPage();
```
[the remark this answers is not in the document]
Comment (jorilindell) [reply src/domain/event/EditEventPage.tsx]: Removed all extra awaits

File: src/common/components/menuDropdown/MenuItem.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    event:
+      | React.FocusEvent<HTMLButtonElement>
+      | React.MouseEvent<HTMLButtonElement>
+  ) => {
+    event.preventDefault();
+    /* istanbul ignore else */
+    if (!disabled) {
+      setFocusedIndex(index);
+    }
+  };
```
[the remark this answers is not in the document]
Comment (jorilindell) [reply src/common/components/menuDropdown/MenuItem.tsx]: Oh, nothing really. I don't know what I have thought. Removed

File: src/common/components/menuDropdown/MenuDropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    polyfill: ResizeObserver,
+  });
+  const disabledIndices = React.useMemo(
+    () =>
+      items.reduce(
+        (acc: number[], item, i) => (item.disabled ? [...acc, i] : acc),
+        []
+      ),
+    [items]
+  );
```
[the remark this answers is not in the document]
Comment (jorilindell) [reply src/common/components/menuDropdown/MenuDropdown.tsx]: You are right, not much benefit to use useMemo here. Removed

**Comment by jorilindell** [comment]
> Screenshot 2021-02-17 at 22 57 31
> 
> Getting this error when running all tests. When I press f to run failed test, it passes

At least that TypeError is coming from hds-react. At some point there was some timer in LoadingSpinned which wasn't cleared on unmount. Not quite sure is that the reason and I'm not able reproduce this locally. 

I have a PR coming where I also update hds-react to latest version. I'm not spending time to this failing test right now because it's passing when running failing tests, but I keep it in my mind.

**Pull request merged**